kafka-cron/producer/main.go

```diff
@@ -0,0 +1,192 @@
+	topicPrefix = kingpin.Flag("topicPrefix", "Topic prefix, e.g. jobs will create topics for each cluster in the format jobs-cluster-a").Default("jobs").String()
+	cronPath    = kingpin.Flag("cronPath", "Path to cron file").Default("./cron.yaml").String()
+	partitions  = kingpin.Flag("partitions", "Number of partitions").Default("1").Int32()
+)
+
+func main() {
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+
+	admin, err := sarama.NewClusterAdmin(*brokerList, config)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+		if err := admin.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	cmds, err := ReadConfig(*cronPath)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = CreateTopics(admin, cmds)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = ScheduleJobs(producer, cmds)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	sigs := make(chan os.Signal, 1)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	<-sigs
+}
+
+func CreateTopics(admin sarama.ClusterAdmin, cmds []types.Command) error {
+	// Create topics for all unique clusters
+	clusters := make([]string, 0)
+
+	for _, cmd := range cmds {
+		for _, cluster := range cmd.Clusters {
+			if !contains(clusters, cluster) {
+				clusters = append(clusters, cluster)
+			}
+		}
+	}
+
+	for _, cluster := range clusters {
+		topicName := fmt.Sprintf("%s-%s", *topicPrefix, cluster)
+
+		err := admin.CreateTopic(topicName, &sarama.TopicDetail{
+			NumPartitions:     *partitions,
+			ReplicationFactor: 1,
+		}, false)
+
+		if err != nil {
+			if errors.Is(err, sarama.ErrTopicAlreadyExists) {
+				admin.CreatePartitions(topicName, *partitions, nil, false)
+			} else {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func ReadConfig(path string) ([]types.Command, error) {
+	var cmds struct {
+		Cron []types.Command `json:"cron" yaml:"cron"`
+	}
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	err = yaml.Unmarshal(data, &cmds)
+	if err != nil {
+		return nil, err
+	}
+	return cmds.Cron, nil
+}
+
+func ScheduleJobs(prod sarama.SyncProducer, cmds []types.Command) error {
+	c := cron.New()
+	for _, cmd := range cmds {
+		fmt.Println("scheduling: ", cmd.Description)
+
+		sch, err := cron.ParseStandard(cmd.Schedule)
+
+		if err != nil {
+			return err
+		}
+
+		job := CommandPublisher{
+			Command:   cmd,
+			publisher: prod,
+		}
+
+		c.Schedule(sch, &job)
+		ScheduledCrons.Inc()
+	}
+	c.Start()
+	return nil
+}
+
+func PublishMessages(prod sarama.SyncProducer, msg string, clusters []string) error {
+	for _, cluster := range clusters {
+		topic := fmt.Sprintf("%s-%s", *topicPrefix, cluster)
```

Comment:
I'd probably extract a function, or even a type, to do this `"%s-%s"` joining behaviour, so that you don't need to update it in multiple places if it changes